src/scanner.rs

```diff
@@ -0,0 +1,400 @@
+use std::collections::HashMap;
+
+use lazy_static::lazy_static;
+use thiserror::Error;
+
+use crate::token::{Literal, Token, TokenType};
+
+lazy_static! {
+    static ref KEYWORDS: HashMap<&'static str, TokenType> = HashMap::from([
+        ("and", TokenType::And),
+        ("class", TokenType::Class),
+        ("else", TokenType::Else),
+        ("false", TokenType::False),
+        ("for", TokenType::For),
+        ("fun", TokenType::Fun),
+        ("if", TokenType::If),
+        ("nil", TokenType::Nil),
+        ("or", TokenType::Or),
+        ("print", TokenType::Print),
+        ("return", TokenType::Return),
+        ("super", TokenType::Super),
+        ("and", TokenType::And),
+        ("this", TokenType::This),
+        ("true", TokenType::True),
+        ("var", TokenType::Var),
+        ("while", TokenType::While),
+    ]);
+}
+
+#[derive(Clone, Error, Debug)]
+#[error("{line}: {message}")]
+pub struct ScannerError {
+    line: usize,
+    message: String,
+}
+
+enum ScanResult<'source> {
+    Skip,
+    Error(ScannerError),
+    Token(Token<'source>),
+}
+
+pub struct Scanner<'source> {
+    source: &'source str,
+    tokens: Vec<Token<'source>>,
+
+    /// First character in the lexeme being scanned
+    start: usize,
```

Comment:
feels like `start` and `current` could be turned into a `&str`, or at least `start` can